### PR TITLE
feat: casing function

### LINKED ISSUE #564: Casing function
**User story.**
As a ruleset maintainer, I would like to have a built-in casing function I could make use of in my own rulesets.
For the time being, in order to enforce casing, you need to use `pattern` function and come up with a regular expression.

**Describe the solution you'd like**
We could make enforcing naming convention in Spectral easier by providing `casing` function with pre-defined set of regular expressions, such as:
- pascal case
- camel case
- kebab case
- snake case
- etc.

`casing` function could also take a custom regular expression, same as `pattern` does.

Here is an example rule definition:
```yaml
given: $.paths[*].[*].[*].operationId # I know it looks weird, it's not important though
then:
      function: casing
      functionOptions:
        case: pascal
```

**Additional context**
It's a complementary functionality to https://github.com/stoplightio/spectral/issues/405

### PR DESCRIPTION
**Pull request by P0lip**
Closes #564

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Additional context**

Open question - what about numbers? Should they be allowed or not? I'd say no yet curious to hear your opinion.

### CODE REVIEW

**Comment by philsturgeon** [comment]
Node Change Case feels like it might have some ideas: https://github.com/blakeembrey/node-change-case

camelCase conversions with numbers in are interesting: 

- from `foo 123-42 bar` to `foo_123_42Bar` 
- from `foo.123.bar` to `foo_123Bar`

seems like we could use the output in their unit tests as validation for our own.

**Comment by P0lip** [comment]
@philsturgeon should we support numbers, though? I excluded them, but feel like we might want them to be included?

Absolutely agree about tests! Thanks for pointing me out at them. Will try to include as many cases as possible.

**Comment by philsturgeon** [comment]
> should we support numbers, though? I

Yes.

**Comment by P0lip** [comment]
> seems like we could use the output in their unit tests as validation for our own.

I took a look and in general there are not many tests I could borrow from them. They aim to support whitespaces, non-ascii characters, and so on. Our case is slightly simpler.
In any case, if you want me to test something specific, let me know and I'll be happy to include such a test case. The more tests the better in this case :)

**Comment by philsturgeon** [comment]
Ahh all i meant with their tests is maybe they have good examples, or you can use it to see what it does. This conversation was about trying to answer "should you support numbers" because basically whatever they'd output is an example of something we'd allow. So, if we make an option for "numbers: true?" then we're probably all set. Or just do whatever is easiest and we can make a followup later.

**Comment by P0lip** [comment]
> So, if we make an option for "numbers: true?" then we're probably all set. Or just do whatever is easiest and we can make a followup later.

I think that's a fair compromise. I'm all in for adding a `numbers` option, as in general I feel like numbers might be desired in some cases, but not really wanted in others, so would be good to let people decide.

**Comment by philsturgeon** [comment]
Yeah let's let numbers through by default, better to be more permissive than too strict here. If people request the option we can add it, but maybe this is right first time. LMK when its rerereview time, I'm excited about this rule!

**Comment by P0lip** [comment]
@philsturgeon added!

File: docs/reference/functions.md [hunk 1]
```diff
@@ -101,6 +101,32 @@ notMatch | if provided, value must _not_ match this regex | no
         notMatch: ".+\/$"
 ```
 
+## casing
+
+Regular expressions!
```
Comment (philsturgeon) [inline docs/reference/functions.md]: Could you explain the 
```suggestion
Text must match a certain case, like `camelCase` or `snake_case`. 
```

File: docs/reference/functions.md [hunk 2]
```diff
@@ -101,6 +101,32 @@ notMatch | if provided, value must _not_ match this regex | no
...
 
+## casing
+
+Regular expressions! 
+
+<!-- title: functionOptions -->
+
+name | description | required?
+---------|----------|---------
+type | the casing type to match against | yes
```
Comment (philsturgeon) [inline docs/reference/functions.md]: What values are acceptable?

File: docs/reference/functions.md [hunk 3]
```diff
@@ -101,6 +101,32 @@ notMatch | if provided, value must _not_ match this regex | no
...
+  description: Name should camelCased.
+  type: style
+  given: "$.name"
+  then:
+    function: casing
+    functionOptions:
+      type: camel
+```
+
+Available types are: flat, camel, pascal, kebab, cobol, snake, macro.
```
Comment (philsturgeon) [inline docs/reference/functions.md]: Oops here they are. Could you move this up near the description and provide examples of each?

**Comment by philsturgeon** [comment]
All good when docs are polished a bit!

File: docs/reference/functions.md (hunk 3 above)
Comment (P0lip) [reply docs/reference/functions.md]: Do you mean to specify valid and invalid examples for each casing type?

File: docs/reference/functions.md [hunk 4]
```diff
@@ -101,9 +101,35 @@ notMatch | if provided, value must _not_ match this regex | no
...
+Text must match a certain case, like `camelCase` or `snake_case`. 
+
+<!-- title: functionOptions -->
+
+name | description | required?
+---------|----------|---------
+type | the casing type to match against | yes
+disallowDigits | if not truthy, digits are allowed | no
+
+Available types are: flat, camel, pascal, kebab, cobol, snake, macro.
```
Comment (nulltoken) [inline docs/reference/functions.md]:
```suggestion
Available types are: 

| name   | sample         |
|--------|----------------|
| flat   | verylongname   |
| camel  | veryLongName   |
| pascal | VeryLongName   |
| kebab  | very-long-name |
| cobol  | VERY-LONG-NAME |
| snake  | very_long_name |
| macro  | VERY_LONG_NAME |
```

Comment (philsturgeon) [reply docs/reference/functions.md]: This! @P0lip accept this one and we can merge.

Comment (P0lip) [reply docs/reference/functions.md]: This is great. Thank you so much @nulltoken

**Pull request merged**